tripy/tripy/frontend/trace/ops/base.py

```diff
@@ -72,21 +73,30 @@ def build(cls, inputs: List["Tensor"], *args, num_outputs=1, **kwargs) -> Union[
         of returning a list of output tensors.
         """
 
-        from tripy.frontend.tensor import Tensor
+        from tripy.common.datatype import int32
         from tripy.frontend.dimension_size import DimensionSize
-
-        # Operations that operate on only DimensionSize inputs will always yield a DimensionSize.
-        # For any mixed operations, DimensionSize must be casted up to Tensor.
-        TensorType = DimensionSize if all(isinstance(inp, DimensionSize) for inp in inputs) else Tensor
+        from tripy.frontend.tensor import Tensor
 
         # NOTE: If you change the stack depth where the tensors are constructed, update STACK_DEPTH_OF_BUILD in
         # the Tensor constructor!
-        outputs = [TensorType(None) for _ in range(num_outputs)]
+        outputs = [Tensor(None) for _ in range(num_outputs)]
 
         inp_trace_tensors = [inp.trace_tensor for inp in inputs]
         out_trace_tensors = [out.trace_tensor for out in outputs]
         cls.build_internal(inp_trace_tensors, out_trace_tensors, *args, **kwargs)
 
+        # Operations that operate on only DimensionSize inputs will always yield a DimensionSize.
+        # For any mixed operations, DimensionSize must be casted up to Tensor.
+        all_inputs_are_dimension_size = all(isinstance(inp, DimensionSize) for inp in inputs)
+        for index, out in enumerate(outputs):
+            if always_cast_to_dimension_size or (
+                all_inputs_are_dimension_size and out.dtype == int32 and out.shape == []
```

Comment:
Will `out.shape` trigger eval here?

Reply:
I think not since these would be trace tensors.

Reply:
Oh good catch, yes it would. I meant to do `out.trace_tensor.shape`

Reply:
Fixed.